src/Chainweb/RestAPI/Backup.hs

```diff
@@ -0,0 +1,90 @@
+{-# LANGUAGE BangPatterns #-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE TypeOperators #-}
+
+-- | An endpoint for making database backups. Also used to synchronize
+-- databases between nodes.
+
+module Chainweb.RestAPI.Backup
+    ( BackupApi
+    , someBackupApi
+    , someBackupServer
+    ) where
+
+import Control.Concurrent.Async
+import Control.Concurrent.STM
+import Control.Monad
+import Control.Monad.Catch
+import Control.Monad.IO.Class
+import Data.Proxy
+import Data.Text (Text)
+import qualified Data.Text as T
+import System.IO.Unsafe
+import System.LogLevel
+
+import Servant
+
+import qualified Chainweb.Backup as Backup
+import Chainweb.Logger
+import Chainweb.Time
+
+import Chainweb.RestAPI.Utils
+import Chainweb.Version
+
+type BackupApi_
+  =    "make-backup" :> QueryFlag "backupPact" :> PostAccepted '[PlainText] Text
+  :<|> "check-backup" :> Capture "backup-name" FilePath :> Get '[PlainText] Backup.BackupStatus
+
+type BackupApi (v :: ChainwebVersionT) = 'ChainwebEndpoint v :> Reassoc BackupApi_
+
+backupApi :: forall (v :: ChainwebVersionT). Proxy (BackupApi v)
+backupApi = Proxy
+
+globalCurrentBackup :: TVar (Maybe Text)
+globalCurrentBackup = unsafePerformIO $! newTVarIO Nothing
+{-# NOINLINE globalCurrentBackup #-}
+
+someBackupApi :: ChainwebVersion -> SomeApi
+someBackupApi (FromSingChainwebVersion (SChainwebVersion :: Sing v)) = SomeApi $ backupApi @v
+
+someBackupServer :: Logger logger => ChainwebVersion -> Backup.BackupEnv logger -> SomeServer
+someBackupServer (FromSingChainwebVersion (SChainwebVersion :: Sing vT)) backupEnv =
+    SomeServer (Proxy @(BackupApi vT)) $ makeBackup :<|> checkBackup
+  where
+    noSuchBackup = err404 { errBody = "no such backup" }
+    makeBackup backupPactFlag = liftIO $ do
+        nextBackupIdentifier <- getNextBackupIdentifier
+        join $ atomically $ do
+            current <- readTVar globalCurrentBackup
+            case current of
+                Nothing -> do
+                    writeTVar globalCurrentBackup (Just nextBackupIdentifier)
+                    return $ doBackup backupPactFlag nextBackupIdentifier
+                Just b -> do
+                    let logg = logFunctionText (Backup._backupLogger backupEnv) Info $
+                            "requested backup, but backup " <> b <> " is already in progress."
+                    return $ b <$ logg
+    doBackup backupPactFlag nextBackupIdentifier = do
+        _ <- async $ do
+            Backup.makeBackup backupEnv options
+            atomically (writeTVar globalCurrentBackup Nothing)
```

Comment:
I think, we need to use `finally` here. Otherwise backups may block infinitely if something goes wrong. Or is that handled elsewhere?

If it is handled elsewhere, consider either moving things around to make the invariants around this variable local, or add a comment that references the respective code locations.

Reply:
It would be safer to run writing of the `TVar` and the call of `doBackup` under a mask and unmask only  `Backup.makeBackup`.

Although, the `join. atomically` pattern seems rather tight. So not sure if we need a mask.

Reply:
> I think, we need to use `finally` here. Otherwise backups may block infinitely if something goes wrong. Or is that handled elsewhere?
> 
> If it is handled elsewhere, consider either moving things around to make the invariants around this variable local, or add a comment that references the respective code locations.

I was thinking this was impossible because `makeBackup` handles all of its errors, but actually there is one case where it doesn't, which is if the backup folder is not writable. I added `finally` so that we can recover from this without restarting the node.